bindings/rust/evmc-vm/src/lib.rs

```diff
@@ -2,6 +2,190 @@ pub extern crate evmc_sys;
 pub use evmc_sys as ffi;
 
 // TODO: Add helpers for host interface
-// TODO: Add convenient helpers for evmc_result
 // TODO: Add convenient helpers for evmc_execute
 // TODO: Add a derive macro here for creating evmc_create
+
+/// EVMC result structure.
+pub struct ExecutionResult {
+    status: ffi::evmc_status_code,
+    gas: i64,
+    output: Option<Vec<u8>>,
+    create_addr: ffi::evmc_address,
+}
+
+impl ExecutionResult {
+    pub fn new(
+        _status: ffi::evmc_status_code,
+        _gas: i64,
+        _output: Option<Vec<u8>>,
+        _create_addr: ffi::evmc_address,
+    ) -> Self {
+        ExecutionResult {
+            status: _status,
+            gas: _gas,
+            output: _output,
+            create_addr: _create_addr,
+        }
+    }
+
+    pub fn get_status(&self) -> ffi::evmc_status_code {
+        self.status
+    }
+
+    pub fn get_gas_left(&self) -> i64 {
+        self.gas
+    }
+
+    pub fn get_output(&self) -> Option<&Vec<u8>> {
+        self.output.as_ref()
+    }
+
+    pub fn get_create_addr(&self) -> Option<&ffi::evmc_address> {
+        // Only return Some if the address is valid (e.g. the status is EVMC_SUCCESS)
+        if self.status == ffi::evmc_status_code::EVMC_SUCCESS {
+            Some(&self.create_addr)
+        } else {
+            None
+        }
+    }
+}
+
+impl From<ffi::evmc_result> for ExecutionResult {
+    fn from(result: ffi::evmc_result) -> Self {
+        let ret = ExecutionResult {
+            status: result.status_code,
+            gas: result.gas_left,
+            output: if !result.output_data.is_null() {
+                // Pre-allocate a vector.
+                let mut buf: Vec<u8> = Vec::with_capacity(result.output_size);
+
+                unsafe {
+                    // Set the len of the vec manually.
+                    buf.set_len(result.output_size);
+                    // Copy from the C struct's buffer to the vec's buffer.
+                    std::ptr::copy(result.output_data, buf.as_mut_ptr(), result.output_size);
+                }
+
+                Some(buf)
+            } else {
+                None
+            },
+            create_addr: result.create_address,
+        };
+
+        // Release allocated ffi struct.
+        unsafe {
+            if result.release.is_some() {
+                result.release.unwrap()(&result as *const ffi::evmc_result);
+            }
+        }
+
+        ret
+    }
+}
+
+/// Returns a pointer to a heap-allocated evmc_result.
+impl Into<*const ffi::evmc_result> for ExecutionResult {
+    fn into(self) -> *const ffi::evmc_result {
+        let (buffer, len) = if let Some(buf) = self.output {
+            let buf_len = buf.len();
+
+            // Manually allocate heap memory for the new home of the output buffer.
+            let memlayout = std::alloc::Layout::from_size_align(buf_len, 1).expect("Bad layout");
+            let new_buf = unsafe { std::alloc::alloc(memlayout) };
```

Comment:
Does `release_result` need to manually deallocate this?

Reply:
yes. I am working on it right now.

Reply:
Isn't it easier to Box it?

Reply:
no, because boxing the pointer to the data will do nothing except kill the pointer. and a box cannot be made over an arbitrary-length buffer.

Reply:
From https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees:
> If a Vec has allocated memory, then the memory it points to is on the heap (as defined by the allocator Rust is configured to use by default), and its pointer points to len initialized, contiguous elements in order (what you would see if you coerced it to a slice), followed by capacity-len logically uninitialized, contiguous elements.

Reply:
`Vec` does not own the data. It owns a pointer to the data on the heap, which is safely allocated and deallocated on instantiation and `drop()` respectively.

You can get a pointer to the buffer this way: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_ptr because you can dereference a Vec to a slice.
However, the slice does not own the data. When the Vec goes out of scope, the slice will as well because `drop` will be called on the Vec. There is no easy way to make the Vec surrender ownership of the data.
From the docs:
> The caller must ensure that the slice outlives the pointer this function returns, or else it will end up pointing to garbage.

Perhaps you could use `mem::transmute` wizardry but this is terribly difficult to reason about.